orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py

```diff
@@ -2054,31 +2054,171 @@ def forward(self, a, b, c, d, *args, kw_0=None, **kwargs):
         assert out.item() == 5.0
         if model.training:
             out.sum().backward()
 
+        # Force model (re)export to validate (un)flattening with new input
+        model._execution_manager(model._is_training())._onnx_model = None
         out = model(one, two, c=three, d=four)
         assert out.item() == 5.0
         if model.training:
             out.sum().backward()
 
+        # Force model (re)export to validate (un)flattening with new input
+        model._execution_manager(model._is_training())._onnx_model = None
         # Model only uses a,d,args[-1] out of a,b,c,d,*args
         out = model(one, two, three, four, *args)
         assert out.item() == 7.0
         if model.training:
             out.sum().backward()
 
+        # Force model (re)export to validate (un)flattening with new input
+        model._execution_manager(model._is_training())._onnx_model = None
         # Model only uses a,d,args[-1],kw_0 out of a,b,c,d,*args,kw_0
         out = model(one, two, three, four, *args, kw_0=kw_0)
         assert out.item() == 13.0
         if model.training:
             out.sum().backward()
 
+        # Force model (re)export to validate (un)flattening with new input
+        model._execution_manager(model._is_training())._onnx_model = None
         # Model only uses a,d,args[-1],kwargs['kwargs_1'] out of a,b,c,d,*args,kw_0,**kwargs
         out = model(one, two, three, four, *args, **kwargs)
         assert out.item() == 15.0
         if model.training:
             out.sum().backward()
+
+
+def test_forward_call_kwargs_input_unexpected_order():
+    class OrderlyNet(torch.nn.Module):
+        def __init__(self, input_size, hidden_size, num_classes):
+            super(OrderlyNet, self).__init__()
+            self.fc1 = torch.nn.Linear(input_size, hidden_size)
+            self.relu = torch.nn.ReLU()
+            self.fc2 = torch.nn.Linear(hidden_size, num_classes)
+
+        def forward(self, input1=None, input2=None):
+            assert input1.shape != input2.shape
+            input2 = torch.transpose(input2, 0, 1)
+            assert input1.shape == input2.shape
+
+            model_input = input1 + input2
+            out1 = self.fc1(model_input)
+            out1 = self.relu(out1)
+            out2 = self.fc2(out1)
+            return out1, out2
+
+    device = 'cuda'
+    N, D_in, H, D_out = 32, 784, 500, 10
+    model = OrderlyNet(D_in, H, D_out).to(device)
+    model = ORTModule(model)
+
+    input1 = torch.randn(N, D_in, device=device, requires_grad=False)
+    input2 = torch.randn(D_in, N, device=device, requires_grad=False)
+
+    # Make sure model runs without any exception
+    for i in range(2):
+        # Test both train and inference mode
+        if i % 2 == 0:
+            model.train()
+        else:
+            model.eval()
+
+        # Force model (re)export to validate (un)flattening with new input
+        model._execution_manager(model._is_training())._onnx_model = None
```

Comment:
Why do you need this (here and elsewhere)? Since this is a private implementation detail of the module, I'd expect the right thing to happen regardless... and maybe you'd want to assert that the re-export happened afterwards?

Reply:
I have added a comment on why we need it and removed some places that we dont strictly need it, but I have added to protect the code from future changes as the issue is subtle

```python
        # Force model (re)export to validate (un)flattening with new input
        #   This is needed because for a `forward(self, a, b)`, and
        #   input `forward(a,b)` or `forward(**{'a': a, 'b': b})`,
        #   ORTModule produces the same schema, thus not re-exporting
        #   the model when `forward(a,b)` is used after `forward(**{'a': a, 'b': b})`
        #   or vice-versa
```